js-scripts/generateRedemptionRights/index.ts

```diff
@@ -0,0 +1,352 @@
+    "0x24C694d193B19119bcDea9D40a3b0bfaFb281E6D": 6487631537430741114,
+    "0x44Cb2d713BDa3858001f038645fD05E23E5DE03D": 27597767454066598826095,
+  },
+};
+
+// Config
+dotenv.config({ path: "./js-scripts/generateRedemptionRights/.env" });
+const config = {
+  rpcUrl: process.env.MAINNET_RPC_URL,
+  pTokenAddresses: (process.env.PTOKEN_ADDRESSES as string)?.split(
+    ","
+  ) as Address[],
+  pointsIds: (process.env.POINTS_IDS as string)?.split(",") as `0x${string}`[],
+  rewardsPerPToken: (process.env.REWARDS_PER_P_TOKEN as string)?.split(
+    ","
+  ) as string[],
+};
+
+// Core functions
+async function calculateRedemptionRights(
+  client: PublicClient,
+  pTokenAddress: Address,
+  rewardsMultiplier: string
+): Promise<[Map<Address, bigint>, PTokenSnapshot]> {
+  const redemptionRights = new Map<Address, bigint>();
+  const rewardsMultiplierBigInt = BigInt(rewardsMultiplier);
+  const blockNumber = await client.getBlockNumber();
+
+  const logs = await client.getLogs({
+    address: pTokenAddress,
+    event: parseAbiItem(
+      "event Transfer(address indexed from, address indexed to, uint256)"
+    ),
+    fromBlock: 0n,
+    toBlock: "latest",
+  });
+
+  // Track raw pToken balances separately
+  const pTokenBalances = new Map<Address, bigint>();
+
+  for (const log of logs) {
+    const [from, to, value] = log.args as [Address, Address, bigint];
+
+    // Update redemption rights
+    if (from !== zeroAddress) {
+      redemptionRights.set(
+        from,
+        (redemptionRights.get(from) || 0n) -
+          (value * rewardsMultiplierBigInt) / BigInt(1e18)
+      );
+      pTokenBalances.set(from, (pTokenBalances.get(from) || 0n) - value);
+    }
+    if (to !== zeroAddress) {
+      redemptionRights.set(
+        to,
+        (redemptionRights.get(to) || 0n) +
+          (value * rewardsMultiplierBigInt) / BigInt(1e18)
+      );
+      pTokenBalances.set(to, (pTokenBalances.get(to) || 0n) + value);
+    }
+  }
+
+  // Apply UNI pool overrides after processing all transfers
+  for (const [poolAddress, redistributions] of Object.entries(
+    UNI_POOL_OVERRIDES
+  )) {
+    const poolBalance = pTokenBalances.get(poolAddress as Address) || 0n;
+    if (poolBalance > 0n) {
+      // Remove balance from pool
+      pTokenBalances.set(poolAddress as Address, 0n);
+      redemptionRights.set(poolAddress as Address, 0n);
+
+      // Redistribute to specified addresses
+      for (const [recipient, amount] of Object.entries(redistributions)) {
+        const recipientAddress = recipient as Address;
+        const overrideAmount = BigInt(amount);
+
+        pTokenBalances.set(
+          recipientAddress,
+          (pTokenBalances.get(recipientAddress) || 0n) + overrideAmount
+        );
+
+        redemptionRights.set(
+          recipientAddress,
+          (redemptionRights.get(recipientAddress) || 0n) +
+            (overrideAmount * rewardsMultiplierBigInt) / BigInt(1e18)
+        );
+      }
+    }
+  }
```

Comment:
I think we need to account for unclaimed pTokens. I think they should essentially be thought of as the unclaimed user's pTokens (even though they arn't claimed). Otherwise unclaimed pTokens at the time of the first snapshot can't be redeemed until all vesting has completed.

Reply:
yes! definitely. I did when calculating `Rewards per p token`, but then missed it here for the redemption rights. I'll add that in (and yea, consider them as balances belonging to the person able to claim them). 

(for sure. also you made me think – we could call claim for all of the users who haven't already in the alpha to make this script work as is hah)